yw_website/yw_website/apps/website/serializers.py

```diff
@@ -1,154 +1,280 @@
-            user=self.context.get('username'),
-            title=validated_data.get('title'),
-            description=validated_data.get('description')
+            user=self.context.get("username"),
+            title=validated_data.get("title"),
+            description=validated_data.get("description"),
         )
         w.save()
         v = Version(
             workflow=w,
-            yw_model_checksum=validated_data.get('modelChecksum'),
-            yw_model_output=validated_data.get('model'),
-            yw_graph_output=validated_data.get('graph'),
-            last_modified=datetime.datetime.now(tz=timezone.utc)
+            yw_model_checksum=validated_data.get("modelChecksum"),
+            yw_model_output=validated_data.get("model"),
+            yw_graph_output=validated_data.get("graph"),
+            last_modified=datetime.datetime.now(tz=timezone.utc),
         )
         v.save()
-        r = self.__create_update_helper(w, v, validated_data)
-
+        r = self._create_update_helper(w, v, validated_data)
         return w.pk, v.pk, r.pk
-    
+
     def update(self, validated_data):
-        w = Workflow.objects.get(pk=self.context.get('workflow_id'))
+        w = Workflow.objects.get(pk=self.context.get("workflow_id"))
         v, new_version = Version.objects.update_or_create(
             workflow=w,
-            yw_model_checksum=validated_data.get('modelChecksum'),
+            yw_model_checksum=validated_data.get("modelChecksum"),
             defaults={
-                'yw_model_output': validated_data.get('model'),
-                'yw_graph_output': validated_data.get('graph'),
-                'last_modified':datetime.datetime.now(tz=timezone.utc)
-            }
+                "yw_model_output": validated_data.get("model"),
+                "yw_graph_output": validated_data.get("graph"),
+                "last_modified": datetime.datetime.now(tz=timezone.utc),
+            },
         )
         v.save()
-        r = self.__create_update_helper(w, v, validated_data)
+        r = self._create_update_helper(w, v, validated_data)
 
         return w.pk, v.pk, r.pk, new_version
-    
-    def __create_update_helper(self, w, v, validated_data):
-        r = self.__create_run(v, validated_data)
-        self.__create_tags(w, validated_data)
-        self.__create_scripts(v, validated_data)
-        self.__create_files(r, validated_data)
+
+    def _create_update_helper(self, w, v, validated_data):
+        r = self._create_run(v, validated_data)
+        self._create_tags(w, validated_data)
+        self._create_scripts(v, validated_data)
+        self._create_files(r, validated_data)
+        self._create_program_blocks(r, validated_data)
+        self._create_ports(r, validated_data)
+        self._create_channels(r, validated_data)
+        self._create_uri_variables(r, validated_data)
+        self._create_resources(r, validated_data)
+        self._create_uri_variable_values(r, validated_data)
```

Comment:
these need to be created in the following order (tags, scripts, and files can be in any order)
- program_blocks
- data
- ports
- channel
- uri_variable
- resource
- uri_variable_value

Otherwise, the serializer will think it is getting bad input when it is getting good input.

Reply:
seems to already be in this order?

Reply:
oh i see, the real problem is that `_create_data` is never called